File: R-Name S/R-Name S_/z.Objective C/StringsManager.m
//
//  StringsManager.m
//  R-Name
//
//  Created by tayo on Tue Apr 02 2002.
//  Copyright (c) 2002 tayo. All rights reserved.
//

#import "StringsManager.h"


@implementation StringsManager

///////// Unavailable Functions for class users //////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark Unavailable_Functions_for_class_users






/////////// Public Methods ///////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark Public_Methods


- (id) init
{
   self = [super init];
   
   bMyCaseSensitive = YES;
   bMyReplaceAllFound = YES;
   
   return self;
}

- (void) dealloc
{
   [super dealloc];
}


// Configuring an StringManager ///////////////////////
- (BOOL) caseSensitive
{
   return bMyCaseSensitive;
}

- (void) setCaseSensitive: (BOOL)bCaseSensitive
{
   bMyCaseSensitive = bCaseSensitive;
}

- (BOOL) replaceAllFound
{
   return bMyReplaceAllFound;
}

- (void) setReplaceAllFound: (BOOL)bReplaceAllFound
{
   bMyReplaceAllFound = bReplaceAllFound;
}

// End Configuring an StringManager ///////////////////





// This method find & replace string.
// This method returns YES if string was found & replaced, otherwise NO.
// For example
// 	strString = @"Hello, World"
// 	strFind = @"Hello"
// 	strReplace = @"Good morning"
//			--> Return YES, strStringModified = @"Good morning, World".
- (BOOL) modifyString: (NSString *)strStringToModify
             toString: (NSString **)strStringModified
                 find: (NSString *)strFind
              replace: (NSString *)strReplace
{
   NSScanner *scScanner;
   NSString *strScaned;
   BOOL bFound = NO;
   NSMutableString *mstrModifiedString;
   
   mstrModifiedString = [NSMutableString stringWithCapacity: [strStringToModify length]];
   
   scScanner = [NSScanner scannerWithString: strStringToModify];
   [scScanner setCaseSensitive: bMyCaseSensitive];
   [scScanner setCharactersToBeSkipped: nil];
   
   // In this case, for example
   // find: "me", replace: "you",   "me and me" --> "you and you"
   if (bMyReplaceAllFound == YES) {
      while(![scScanner isAtEnd]) {
         if ([scScanner scanUpToString: strFind intoString: &strScaned]) {
            [mstrModifiedString appendString: strScaned];
         }
         if ([scScanner scanString: strFind intoString: &strScaned]) {
            [mstrModifiedString appendString: strReplace];
            bFound = YES;
         }
      }
   }
   // In this case, for example
   // find: "me", replace: "you",   "me and me" --> "you and me"
   else { // bMyReplaceAllFound == NO
      while(![scScanner isAtEnd]) {
         if ([scScanner scanUpToString:strFind intoString: &strScaned]) {
            [mstrModifiedString appendString: strScaned];
         }
         if ([scScanner scanString:strFind intoString:nil]) {
            if (bFound == NO) {
               [mstrModifiedString appendString: strReplace];
            }
            else {
               [mstrModifiedString appendString: strFind];            
            }
            
            bFound = YES;      
         }
      }
   }
   
   *strStringModified = mstrModifiedString;
   return bFound;
}





// This method returns string that is converted '/' and ':'
- (NSString *) stringWithSlushAndColonConverted: (NSString *)strBefore
{
   NSMutableString *strAfter = [NSMutableString stringWithCapacity: [strBefore length]];
   NSScanner *scScanner = [NSScanner scannerWithString: strBefore];
   NSCharacterSet *csSlushOrColon = [NSCharacterSet characterSetWithCharactersInString: @"/:"];
   NSString *strNoSlushOrColon, *strSlushOrColon;

   if (strBefore == nil) {
      return nil;
   }
   if ([strBefore length] == 0) {
      return @"";
   }

   [scScanner setCharactersToBeSkipped: nil];
   while(![scScanner isAtEnd]) {
      if ([scScanner scanUpToCharactersFromSet: csSlushOrColon intoString: &strNoSlushOrColon]) {
         [strAfter appendString: strNoSlushOrColon];
      }
      if ([scScanner scanCharactersFromSet: csSlushOrColon intoString: &strSlushOrColon]) {
         int iCount;

         for (iCount = 0; iCount < [strSlushOrColon length]; iCount++) {
            if ([strSlushOrColon characterAtIndex: iCount] == '/') {
               [strAfter appendString: @":"];
            }
            else { // if it is ':'
               [strAfter appendString: @"/"];
            }
         }
      }
   }

   return strAfter;
}




@end

File: R-Name S/Stubs/StringsManager.swift
//
//  StringsManager.swift
//  R-Name S
//
//  Created by Pedro Plowman on 13/5/19.
//  Copyright © 2019 p2made. All rights reserved.
//

import Foundation

class StringsManager: NSObject {
	var defaultCaseSensitive = true // default true;
	var defaultReplaceAllFound = true // default true;

	// Configuring StringManager
	func caseSensitive() -> Bool {
		return defaultCaseSensitive
	}

	func setCaseSensitive(_ bCaseSensitive: Bool) {
		defaultCaseSensitive = bCaseSensitive
	}

	func replaceAllFound() -> Bool {
		return defaultReplaceAllFound
	}

	func setReplaceAllFound(_ bReplaceAllFound: Bool) {
		defaultReplaceAllFound = bReplaceAllFound
	}

	// This method find & replace string.
	// This method returns YES if string was found & replaced, otherwise NO.
	// For example
	// 	strString = @"Hello, World"
	// 	strFind = @"Hello"
	// 	strReplace = @"Good morning"
	//			--> Return YES, strStringModified = @"Good morning, World".
	func modifyString(_ strStringToModify: String?, to strStringModified: String?, find strFind: String?, replace strReplace: String?) -> Bool {
		var strStringModified = strStringModified
		var scScanner: Scanner?
		var strScaned: String
		var bFound = false
		var mstrModifiedString: String

		mstrModifiedString = String(repeating: "\0", count: strStringToModify?.count ?? 0)

		scScanner = Scanner(string: strStringToModify ?? "")
		scScanner?.caseSensitive = caseSensitive()
		scScanner?.charactersToBeSkipped = nil

		// In this case, for example
		// find: "me", replace: "you",   "me and me" --> "you and you"
		if replaceAllFound() {
			while !scScanner?.isAtEnd() {
				if scScanner?.scanUpTo(strFind ?? "", into: &strScaned) ?? false {
					mstrModifiedString += strScaned
				}
				if scScanner?.scanString(strFind ?? "", into: &strScaned) ?? false {
					mstrModifiedString += strReplace ?? ""
					bFound = true
				}
			}
		} else {
			// bMyReplaceAllFound == NO
			while !scScanner?.isAtEnd() {
				if scScanner?.scanUpTo(strFind ?? "", into: &strScaned) ?? false {
					mstrModifiedString += strScaned
				}
				if scScanner?.scanString(strFind ?? "", into: nil) ?? false {
					if bFound == false {
						mstrModifiedString += strReplace ?? ""
					} else {
						mstrModifiedString += strFind ?? ""
					}

					bFound = true
				}
			}
		}

		strStringModified = mstrModifiedString
		return bFound
	}

	// This method returns string that is converted '/' and ':'
	func string(withSlushAndColonConverted strBefore: String?) -> String? {
		var strAfter = String(repeating: "\0", count: strBefore?.count ?? 0)
		let scScanner = Scanner(string: strBefore ?? "")
		let csSlushOrColon = CharacterSet(charactersIn: "/:")
		var strNoSlushOrColon: String
		var strSlushOrColon: String

		if strBefore == nil {
			return nil
		}
		if (strBefore?.count ?? 0) == 0 {
			return ""
		}

		scScanner.charactersToBeSkipped = nil
		while !scScanner.isAtEnd() {
			if scScanner.scanUpToCharacters(from: csSlushOrColon, into: &strNoSlushOrColon) {
				strAfter += strNoSlushOrColon
			}
			if scScanner.scanCharacters(from: csSlushOrColon, into: &strSlushOrColon) {
				var iCount: Int

				for iCount in 0..<strSlushOrColon.count {
					if strSlushOrColon[strSlushOrColon.index(strSlushOrColon.startIndex, offsetBy: UInt(iCount))] == "/" {
						strAfter += ":"
					} else {
						// if it is ':'
						strAfter += "/"
					}
				}
			}
		}

		return strAfter
	}
}
